java-checks/src/main/java/org/sonar/java/checks/AnonymousClassShouldBeLambdaCheck.java

```diff
@@ -76,14 +78,44 @@ public void visitNewClass(NewClassTree tree) {
   }
 
   private static boolean isSAM(ClassTree classBody) {
-    if(hasOnlyOneMethod(classBody.members()) && classBody.symbol().isTypeSymbol()) {
+    if (hasOnlyOneMethod(classBody.members()) && classBody.symbol().isTypeSymbol()) {
       // Verify class body is a subtype of an interface
+      MethodJavaSymbol methodSymbol = ((MethodJavaSymbol) ((MethodTree) classBody.symbol().memberSymbols().stream().findFirst().get().declaration())
+        .symbol());
       JavaSymbol.TypeJavaSymbol symbol = (JavaSymbol.TypeJavaSymbol) classBody.symbol();
-      return symbol.getInterfaces().size() == 1 && symbol.getSuperclass().is("java.lang.Object");
+      boolean checkAnonymousClassFP = defaultMehodsSameSign(methodSymbol);
+      return symbol.getInterfaces().size() == 1 &&
+        symbol.getSuperclass().is("java.lang.Object") &&
+        checkAnonymousClassFP;
     }
     return false;
   }
 
+  // seaches in the interface if there is a default method with the same signature
+  private static boolean defaultMehodsSameSign(MethodJavaSymbol methodSymbol) {
+    TypeJavaSymbol tjs = methodSymbol.overriddenSymbol().outermostClass();
+    if (methodSymbol.overriddenSymbol().isDefault() && tjs.isInterface() && tjs.declaration() != null) {
+      for (Symbol mjs : ((TypeJavaSymbol) tjs.declaration().symbol()).memberSymbols()) {
+        if (mjs.isMethodSymbol() && !checkmethodSignature(methodSymbol, (MethodJavaSymbol) mjs)) {
+          return false;
+        }
+      }
+    }
+
+    return true;
+  }
+
+  private static boolean checkmethodSignature(MethodJavaSymbol methodSymbol, MethodJavaSymbol met) {
+    if (!methodSymbol.name().equals(met.name()) && met.isDefault()) {
```

Comment:
This is not the correct way to check for equivalence. You don't want to rely on string transformations.